styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -126,37 +136,120 @@
+                                    .withNewMetadata()
+                                    .withName(secretName)
+                                    .withAnnotations(annotations)
+                                    .endMetadata()
+                                    .withData(keys)
+                                    .build());
+        LOG.info("[AUDIT] Secret {} created for service account {}", secretName, serviceAccount);
+      }
+    } else if (runSpec.secret().isPresent()) {
+      final WorkflowConfiguration.Secret specSecret = runSpec.secret().get();
+      final Secret secret = client.secrets().withName(specSecret.name()).get();
+      if (secret == null) {
+        LOG.error("[AUDIT] Workflow {} refers to a non-existent secret {}",
+                  workflowInstance.workflowId(), specSecret.name());
+        throw new InvalidExecutionException(
+            "Referenced secret '" + specSecret.name() + "' was not found");
+      } else {
+        LOG.info("[AUDIT] Workflow {} refers to secret {}",
+                 workflowInstance.workflowId(), specSecret.name());
+      }
+    }
+  }
+
+  private static String buildSecretName(String serviceAccount) {
+    return STYX_WORKFLOW_SA_SECRET_NAME + '-' + Hashing
+        .sha256().hashString(serviceAccount, StandardCharsets.UTF_8);
+  }
+
   @VisibleForTesting
   static Pod createPod(WorkflowInstance workflowInstance, RunSpec runSpec) {
     final String imageWithTag = runSpec.imageName().contains(":")
         ? runSpec.imageName()
         : runSpec.imageName() + ":latest";
 
     final String podName = STYX_RUN + "-" + UUID.randomUUID().toString();
+    final List<EnvVar> env = buildEnv(workflowInstance, runSpec, podName);
+
+    final PodBuilder podBuilder = new PodBuilder()
+        .withNewMetadata()
+        .withName(podName)
+        .addToAnnotations(STYX_WORKFLOW_INSTANCE_ANNOTATION, workflowInstance.toKey())
+        .addToAnnotations(DOCKER_TERMINATION_LOGGING_ANNOTATION, String.valueOf(runSpec.terminationLogging()))
+        .endMetadata();
+    PodFluent.SpecNested<PodBuilder> spec = podBuilder.withNewSpec()
+        .withRestartPolicy("Never");
+    PodSpecFluent.ContainersNested<PodFluent.SpecNested<PodBuilder>> container = spec
+        .addNewContainer()
+            .withName(STYX_RUN)
+            .withImage(imageWithTag)
+            .withArgs(runSpec.args())
+            .withEnv(env);
+
+    if (runSpec.serviceAccount().isPresent()) {
+      spec = spec.addNewVolume()
+          .withName(STYX_WORKFLOW_SA_SECRET_NAME)
+          .withNewSecret()
+          .withSecretName(buildSecretName(runSpec.serviceAccount().get()))
+          .endSecret()
+          .endVolume();
+      container = container
+          .addToVolumeMounts(new VolumeMount(STYX_MANAGED_WORKFLOW_SA_SECRET_MOUNT_PATH,
+                                             STYX_WORKFLOW_SA_SECRET_NAME, true))
+          // TODO: do we need set this env as default value?
```

Comment:
👍 

This way we can migrate to network (metadata server) based SA credentials in the future without the user having to care.